pkg/coordinator/chain.go

```diff
@@ -1,17 +1,69 @@
 package coordinator
 
 import (
+	"math/big"
+
 	"github.com/keep-network/keep-core/pkg/bitcoin"
 	"github.com/keep-network/keep-core/pkg/tbtc"
-	"math/big"
 )
 
+// NewWalletRegisteredEvent represents a new wallet registered event.
+type NewWalletRegisteredEvent struct {
+	EcdsaWalletID       [32]byte
+	WalletPublicKeyHash [20]byte
+	BlockNumber         uint64
+}
+
+// NewWalletRegisteredEventFilter is a component allowing to filter NewWalletRegisteredEvent.
+type NewWalletRegisteredEventFilter struct {
+	StartBlock          uint64
+	EndBlock            *uint64
+	EcdsaWalletID       [][32]byte
+	WalletPublicKeyHash [][20]byte
+}
+
 // Chain represents the interface that the coordinator module expects to interact
 // with the anchoring blockchain on.
 type Chain interface {
-	// TODO: Change to something more specific once https://github.com/keep-network/keep-core/issues/3632
-	// is handled.
-	tbtc.Chain
+	// GetDepositRequest gets the on-chain deposit request for the given
+	// funding transaction hash and output index. Returns an error if the
+	// deposit was not found.
+	GetDepositRequest(
+		fundingTxHash bitcoin.Hash,
+		fundingOutputIndex uint32,
+	) (*tbtc.DepositChainRequest, error)
```

Comment:
As for the functions, it's clear that we want to define them separately for each package, even if it means duplication. But what about the types the functions are expecting or returning?
I don't think duplication is the right way. Alternatively to the current solution where we just import the types from the `tbtc` package, we could alias them, but it doesn't make much sense.

I'm curious about your opinion @pdyraga @lukasz-zimnoch.

Reply:
In my opinion, we should look at the packages as some bounded contexts of the same domain:
- `pkg/tbtc` is the context corresponding to the tbtc wallet
- `pkg/coordinator` is the context of a tbtc wallet coordinator

That said, both should define domain objects (i.e. types) and functions specific to their contexts. However, some of the domain may be shared between those two contexts and this is the case we observe here. I think we have two options:
1. Declare the shared objects in one of the contexts
2. Extract the shared objects to a separate place

I think option 2 is worse in our case as it will introduce unnecessary confusion. In my opinion, we should aim for option 1 as it is more "natural" for our codebase. 

The only question is: "Which package should hold the shared objects?". In our case, the `pkg/tbtc` is kind of a core package while the `pkg/coordinator` is kind of an overlay specialized for the coordination part. If we agree about that, this implies that all shared objects should live in the `pkg/tbtc` package.

Reply:
I would be fine with the duplication for now but I think the problem with the duplication is that those structs are separate types and the chain implementation struct would have to redeclare the same function multiple times. For example:
```
func GetDepositRequest(...) coordinator.DepositChainRequest // pkg/coordinator chain interface
func GetDepositRequest(...) spv.DepositChainRequest // pkg/maintainer/spv chain interface
func GetDepositRequest(...) wallet.DepositChainRequest // pkg/maintainer/wallet chain interface
```

The current solution is the lesser evil but I think we can do better. Instead of creating a dependency to `tbtc` package, we could have a separate package with the common type declarations. 

```
pkg/chain/types/types.go

pkg/tbtc/chain.go                           // imports pkg/chain/types
pkg/coordinator/chain.go                    // imports pkg/chain/types
pkg/maintainer/wallet/chain.go              // imports pkg/chain/types
pkg/maintainer/spv/chain.go                 // imports pkg/chain/types
```

Reply:
As an alternative, we could also use a similar pattern that we use in `pkg/tecdsa`, i.e.:
```
pkg
    tbtc
        wallet           // holds the current contents of pkg/tbtc
        coordinator // holds the current contents of pkg/coordinator
        maintainer   // holds the current contents of pkg/maintainer
        chain.go      // holds interfaces and objects shared by the above packages
```

Reply:
I think we intentionally decided to flatten packages compared to `pkg/tecdsa` and treat `pkg/tbtc` as the client-side wallet implementation. I don't think the alternative is bad but I also like the fact the packages are currently more flat compared to `pkg/tecdsa`.

Reply:
Yeah, that's true and I'm also strongly for the flat packages approach. I just think that if we want to extract the shared part somewhere, we should rather use this pattern as it fits more our domain-oriented package structure than the proposed `/pkg/chain/tbtc/types/types.go` which feels a bit artificial.

Reply:
Anyway, I think the best we can do for now is just keep the shared objects in `pkg/tbtc` and decouple the interfaces. This still makes `pkg/coordinator` dependent on `pkg/tbtc` but we achieve more specialized interfaces and shorter mocks. 

The dependency on `pkg/tbtc` objects is not the end of the world. Note that the existence of that dependency may indicate that the boundaries between our packages are not optimal but it's really hard to achieve the right shape on the first try and we should rather seek gradual improvements over time instead of making a revolution now 🤷

Reply:
> Note that the existence of that dependency may indicate that the boundaries between our packages are not optimal but it's really hard to achieve the right shape on the first try and we should rather seek gradual improvements over time instead of making a revolution now 🤷

I wouldn't be too hard on ourselves. I like the current setup of our packages. It is really clean and separates boundaries well. `pkg/tbtc` is the core responsibility of tBTC client: being a wallet, responding to on-chain requests, and signing stuff. `pkg/coordinator` and `pkg/maintainer` could as well live outside of `keep-network/keep-core` but for a ton of other valid reasons we decided to keep them together in one binary. This is fine. The core of the logic takes the central seat in `pkg/tbtc` and all the auxiliary stuff sits out of the center of the room in case we ever need to move it outside.

Now, there are some common types both for the core logic and the auxiliary logic. `DepositChainRequest` is one of them. It is needed by the core wallet signing logic and the auxiliary maintainer logic. If `pkg/tbtc` is about the wallet signing logic, the auxiliary maintainer logic shouldn't import it. `pkg/tbtc/types` is a completely separate package from the Go perspective. 

I do agree it feels a bit artificial from the domain perspective but I think it does not feel artificial from the Go side and there are some examples of using this pattern in `go-ethereum` as well: [`core/types`](https://github.com/ethereum/go-ethereum/tree/d73eb87979021e63cb5867319c8ef07a9ba8a7d2/core/types) or [`beacon/types`](https://github.com/ethereum/go-ethereum/tree/d73eb87979021e63cb5867319c8ef07a9ba8a7d2/beacon/types).

My slight preference is on `pkg/tbtc/types` package but I could as well have the import to `pkg/tbtc`. I would avoid duplication given the problems in the implementation structure and I would avoid moving the `maintainer` and `coordinator` packages to `pkt/tbtc` because I really like the flat structure and the current way of thinking about the responsibilities of our packages.

Reply:
> If pkg/tbtc is about the wallet signing logic, the auxiliary maintainer logic shouldn't import it. pkg/tbtc/types is a completely separate package from the Go perspective.

Agreed but in fact, `pkg/tbtc` is a little bit broader than we assume here. It also contains some logic that is useful for auxiliary packages, for example `tbtc.ValidateDepositSweepProposal`. That means that even if we extract types to `pkg/tbtc/types`, we would still have the dependency due to some shared logic. This is why I'm leaning towards leaving shared types in `pkg/tbtc` and do not complicate the situation with an additional package.

Reply:
In order to elaborate more on why I'm a bit against `pkg/tbtc/types`:
- It's a bit too generic and does not give a clear answer about what should live inside and what not
- It is nested within `pkg/tbtc` which will make a wrong dependency direction: `pkg/tbtc` -> `pkg/tbtc/types`. In Go, we typically make sub-packages depending on the outer packages (a lot of examples in Go std lib but also ours `pkg/net`, `pkg/tecdsa`, and so on).

If we really insist on extracting, I would rather go with creating `pkg/tbtccore` and put there all core types and shared logic that does not belong to any specific domain context.

Reply:
> * It is nested within `pkg/tbtc` which will make a wrong dependency direction: `pkg/tbtc` -> `pkg/tbtc/types`. In Go, we typically make sub-packages depending on the outer packages (a lot of examples in Go std lib but also ours `pkg/net`, `pkg/tecdsa`, and so on).

Yes, we would have to bend this rule. It is not ideal but we wouldn't be the only ones. For example, the `Block` structure is defined in `github.com/ethereum/go-ethereum/core/types` package and it is imported by `go-ethereum/core/blockchain_insert.go` [here](https://github.com/ethereum/go-ethereum/blob/d73eb87979021e63cb5867319c8ef07a9ba8a7d2/core/blockchain_insert.go#L24). This is exactly our use case. As long as we do it only for the `types` package, I would be personally fine with it. 

> * It's a bit too generic and does not give a clear answer about what should live inside and what not

It is less generic than `tbtccore` :) I see the reasoning behind the `tbtccore` but I am afraid this will sooner or later evolve into a common bag of stuff and create even more complicates mesh of dependencies. I think it's nothing wrong if `coordinator` or `maintainer` packages imports logic from `tbtc`. It is clear, they perform the same validation as the wallet as in the mentioned `tbtc.ValidateDepositSweepProposal` example and in my opinion, this is a really nice feature of the current code organization.

My slight preference is for a separate types package with good documentation of why it's there and what types should be there. This would give us more flexibility and avoid import cycles. But I am also fine with leaving it as-is with imports from `tbtc`.

Reply:
Unfortunately, it seems that there is no perfect solution.

If we decide to declare the types in a common types package what types should go there? All the types used by chain interfaces across all the packages? We have types that are common to `pkg/tbtc` and `pkg/coordinator` and specific to the `pkg/tbtc` only. Which do we want to move to the common types?

Separating types from the chain interfaces declaration may be confusing, as in V2 we decided to declare the chain interfaces closer to the actual usage in packages.

Importing types defined in `pkg/tbtc` package into the `pkg/coordinator` doesn't seem perfect either. The `pkg/coordinator` package uses `pkg/tbt` types, but also declares its' own types (e.g. `NewWalletRegisteredEvent`).

We cannot duplicate the types either, as the functions defined in different packages will require the same type, e.g. `ValidateRedemptionProposal` defined in both `pkg/coordinator` and `pkg/tbtc` must use the same `RedemptionProposal` type.

I lean towards the current approach where types are defined in `pkg/tbtc` which is considered a broader and core package and we just import them in `pkg/coordinator`. This isn't perfect but it doesn't require an aggressive refactor. 

We could add another issue for the future to find the perfect solution for types definition.

Reply:
Captured it here: https://github.com/keep-network/keep-core/issues/3652